File: leetcode/python/199_binary_tree_right_side_view.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def rightSideView(self, root: TreeNode | None) -> list[int]:
        if root is None:
            return []
        
        queue = deque([root])
        res: list[int] = []

        while queue:
            tmp = deque()
            node: TreeNode | None = None
            while queue:
                node = queue.popleft()
                if node.left:
                    tmp.append(node.left)
                if node.right:
                    tmp.append(node.right)
            res.append(node.val)
            queue = tmp
        return res

File: leetcode/c++/199_binary_tree_right_side_view.cpp
#include <vector>
#include <queue>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if (!root) return {};

        vector<int> res;
        queue<TreeNode*> q({root});
        while (!q.empty()) {
            TreeNode* last;
            auto size = q.size();
            for (int i = 0; i < size; ++i) {
                last = q.front();
                q.pop();
                if (last->left) q.push(last->left);
                if (last->right) q.push(last->right);
            }
            res.push_back(last->val);
        }
        return res;
    }
};